lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1997,6 +1997,177 @@ bool swift::diagnoseArgumentLabelError(TypeChecker &TC, const Expr *expr,
+  auto *CE = dyn_cast<CallExpr>(initExpr);
+  if (!CE)
+    return;
+
+  auto *CRCE = dyn_cast<ConstructorRefCallExpr>(CE->getFn());
+  if (!CRCE)
+    return;
+
+  auto *DRE = dyn_cast<DeclRefExpr>(CRCE->getFn());
+  if (!DRE)
+    return;
+
+  auto *constructorDecl = dyn_cast<ConstructorDecl>(DRE->getDecl());
+  if (!constructorDecl)
+    return;
+
+  // Make sure the constructor constructs an instance that allows ownership.
+  // This is to ensure we don't diagnose on constructors such as
+  // Optional.init(nilLiteral:).
+  auto selfType = constructorDecl->getDeclContext()->getSelfTypeInContext();
+  if (!selfType->allowsOwnership())
+    return;
+
+  // This must stay in sync with
+  // diag::unowned_assignment_immediate_deallocation.
+  enum {
+    SK_Variable = 0,
+    SK_Property
+  } storageKind = SK_Variable;
+
+  if (varDecl->getDeclContext()->isTypeContext())
+    storageKind = SK_Property;
+
+  TC.diagnose(initExpr->getStartLoc(),
+              diag::unowned_assignment_immediate_deallocation,
+              varDecl->getName(),
+              (unsigned) ownershipAttr->get(), (unsigned) storageKind)
+    .highlight(initExpr->getSourceRange());
+
+  TC.diagnose(initExpr->getStartLoc(),
+              diag::unowned_assignment_requires_strong)
+    .highlight(initExpr->getSourceRange());
+
+  TC.diagnose(varDecl->getStartLoc(), diag::decl_declared_here,
+              varDecl->getFullName())
+    .highlight(varDecl->getSourceRange());
+}
+
+void swift::diagnoseUnownedImmediateDeallocation(Expr *destExpr,
+                                                 Expr *initExpr,
+                                                 TypeChecker &TC) {
+  destExpr = destExpr->getValueProvidingExpr();
+
+  // Try to find a referenced VarDecl.
+  VarDecl *VD = nullptr;
+  if (auto *DRE = dyn_cast<DeclRefExpr>(destExpr)) {
+    VD = dyn_cast<VarDecl>(DRE->getDecl());
+  } else if (auto *MRE = dyn_cast<MemberRefExpr>(destExpr)) {
+    VD = dyn_cast<VarDecl>(MRE->getMember().getDecl());
+  }
+
+  if (VD)
+    diagnoseUnownedImmediateDeallocationImpl(VD, initExpr, TC);
+}
+
+void swift::diagnoseUnownedImmediateDeallocation(Pattern *pattern,
+                                                 Expr *initExpr,
+                                                 TypeChecker &TC) {
+  pattern = pattern->getSemanticsProvidingPattern();
+
+  if (auto *TP = dyn_cast<TuplePattern>(pattern)) {
+    initExpr = lookThroughExprsToImmediateDeallocation(initExpr);
+
+    // If we've found a matching tuple initializer with the same number of
+    // elements as our pattern, diagnose each element individually.
+    auto TE = dyn_cast<TupleExpr>(initExpr);
+    if (TE && TE->getNumElements() == TP->getNumElements()) {
+      for (unsigned i = 0, e = TP->getNumElements(); i != e; ++i) {
+        TuplePatternElt &elt = TP->getElement(i);
+        Pattern *subPattern = elt.getPattern();
+        Expr *subInitExpr = TE->getElement(i);
```

Comment:
You should `lookThroughExprsToImmediateDeallocation` again here, no?

Reply:
@jrose-apple We don't have to; if it's a nested `TuplePattern`, then we `lookThroughExprsToImmediateDeallocation` on its `initExpr` (in the recursive call) before digging into the elements, and if it's a nested `NamedPattern` then `diagnoseUnownedImmediateDeallocationImpl` already calls it.

Reply:
Ah, I missed that this was a recursive call and not directly calling the `Impl` variant. My mistake!